### PR TITLE
Add CniServer into HostDaemon

### PR DESCRIPTION
**Pull request by bn222**
This PR adds functionality to propagate a CNI add for a device that is added on the host down to the DPU which in turns forwards that to the vendor plugin on the DPU.

### CODE REVIEW

File: daemon/main.go [hunk 1]
```diff
@@ -53,6 +80,12 @@ func main() {
 	log := ctrl.Log.WithName("Daemon Init")
 	log.Info("Daemon init")
 
+	err = copyCNI()
```
Comment (wizhaoredhat) [inline daemon/main.go:83]: Unfortunately due to the delta between RHEL versions we should be using the `entrypoint.sh` method to copy the CNI. Same code for SRIO-CNI and Multus:

https://github.com/openshift/sriov-cni/pull/100#issuecomment-2009001716

File: go.mod [hunk 2]
```diff
@@ -21,7 +21,7 @@ require (
 	go.uber.org/zap v1.26.0
 	golang.org/x/net v0.19.0
 	google.golang.org/grpc v1.58.3
-	google.golang.org/protobuf v1.31.0
+	google.golang.org/protobuf v1.33.0
```
Comment (wizhaoredhat) [inline go.mod:24]: Is this upgrade to utilize new APIs? Maybe a helpful commit message may make it easier to for people to know happened here.

File: daemon/hostdaemon.go [hunk 3]
```diff
@@ -89,4 +99,20 @@ func (d *HostDaemon) Start() {
...
+	server := cniserver.NewCNIServer()
+	err = server.Start()
+
+	if err != nil {
+		d.log.Error(err, "Error starting CNI server for shim")
+	}
+
+	d.cniserver = server
+	go func() {	
+		d.cniserver.Start()
```
Comment (wizhaoredhat) [inline daemon/hostdaemon.go:168]: This code is kind of confusing. We do server.Start() above and we are doing the same server.Start() in a Go routine.

File: daemon/main_test.go [hunk 4]
```diff
@@ -27,18 +33,61 @@ var _ = g.BeforeSuite(func() {
...
 var _ = g.Describe("Main", func() {
 	g.Context("Connection", func() {
-		dummyPluginDPU  := plugin.NewDummyPlugin()
+		dummyPluginDPU := plugin.NewDummyPlugin()
 		DpuDaemon := NewDpuDaemon(dummyPluginDPU)
 
 		dummyPluginHost := plugin.NewDummyPlugin()
 		HostDaemon := NewHostDaemon(dummyPluginHost)
 
+		g.It("Should connect succesfully if the server is up first", func() {
```
Comment (wizhaoredhat) [inline daemon/main_test.go:64]: "Should connect successfully if the DPU Daemon is up first"

File: daemon/main_test.go [hunk 5]
```diff
@@ -27,18 +33,61 @@ var _ = g.BeforeSuite(func() {
...
+			DpuDaemon.Start()
+			HostDaemon.Start()
+			err := HostDaemon.CreateBridgePort(1,1,1)
+			Expect(err).ShouldNot(HaveOccurred())
+
+			HostDaemon.Stop()
+			DpuDaemon.Stop()
+		})
 		g.It("Should connect succesfully if the server is up first", func() {
```
Comment (wizhaoredhat) [inline daemon/main_test.go:73]: "Should connect successfully if the Host Daemon is up first"

File: dpu-cni/pkgs/cniserver/cniserver.go [hunk 6]
```diff
@@ -247,14 +248,23 @@ func (s *Server) handleCNIRequest(r *http.Request) ([]byte, error) {
...
 	if err != nil {
 		return nil, err
 	}
+	defer req.Cancel()
 
-	response := &cnitypes.Response{}
-	response.Result = result
+	var result *cni100.Result = nil
+	if req.Command == cnitypes.CNIAdd {
+		result, err = s.addHandler(req)
```
Comment (wizhaoredhat) [inline dpu-cni/pkgs/cniserver/cniserver.go:262]: Can we rename "addHandler" to  "cniCmdAddHandler"... "addHandler" and "delHandler" has different meanings on its own.

File: dpu-cni/pkgs/cniserver/cniserver_test.go [hunk 7]
```diff
@@ -59,13 +42,35 @@ var _ = g.Describe("Cniserver", func() {
...
 	var err error
+	var addHandlerCalled bool = false
+	var delHandlerCalled bool = false
+
+	addHandler := func(request *cnitypes.PodRequest) (*current.Result, error) {
+		result := &current.Result{
+			CNIVersion: request.CNIConf.CNIVersion,
+		}
+
+		addHandlerCalled = true
```
Comment (wizhaoredhat) [inline dpu-cni/pkgs/cniserver/cniserver_test.go:53]: Let's rename "addHandler*" to something like "cniCmdAddHandler*"

File: dpu-cni/pkgs/cniserver/cniserver_test.go [hunk 8]
```diff
@@ -80,11 +85,26 @@ var _ = g.Describe("Cniserver", func() {
 				CNIVersion: cniVersion,
 			}
 			g.It("should get a correct response from the post request", func() {
-				resp, ver, err := plugin.PostRequest(PrepArgs(cniVersion))
+				resp, ver, err := plugin.PostRequest(PrepArgs(cniVersion, "ADD"))
```
Comment (wizhaoredhat) [inline dpu-cni/pkgs/cniserver/cniserver_test.go:86]: (Optional) I know it is horrible on its on. But can we have a const added something for ADD and DEL? like "cnitypes.CommandAdd"

File: daemon/hostdaemon.go [hunk 9]
```diff
@@ -26,11 +29,11 @@ type HostDaemon struct {
 	cniserver *cniserver.Server
 }
 
-func (d *HostDaemon) CreateBridgePort(pf int, vf int, vlan int) (error) {
+func (d *HostDaemon) CreateBridgePort(pf int, vf int, vlan int, mac string) (error) {
```
Comment (wizhaoredhat) [inline daemon/hostdaemon.go:33]: (Comment) Remember how we might want to have a json for the name... We are fast approaching needing this.

File: daemon/hostdaemon.go [hunk 10]
```diff
@@ -90,6 +93,36 @@ func (d *HostDaemon) ensureConnected() {
...
+	pf := 0
+	vf := req.CNIConf.VFID
+	mac := req.CNIConf.MAC
+	vlan := 7
+	d.CreateBridgePort(pf, vf, vlan, mac)
+
+	sm := sriov.NewSriovManager()
+	res, err := sm.CmdAdd(req)
+	if err != nil {
+	    return nil, errors.New("SRIOV manager falied in add handler")
```
Comment (wizhaoredhat) [inline daemon/hostdaemon.go]: SP: Failed

File: daemon/hostdaemon.go [hunk 11]
```diff
@@ -90,6 +93,36 @@ func (d *HostDaemon) ensureConnected() {
...
+	pf := 0
+	vf := req.CNIConf.VFID
+	mac := req.CNIConf.MAC
+	vlan := 7
+	d.DeleteBridgePort(pf, vf, vlan, mac)
+
+	sm := sriov.NewSriovManager()
+	err := sm.CmdDel(req)
+	if err != nil {
+	    return nil, errors.New("SRIOV manager falied in del handler")
```
Comment (wizhaoredhat) [inline daemon/hostdaemon.go:137]: SP: Failed

**Review by wizhaoredhat — CHANGES_REQUESTED** [review]
Part 1 of review stopping at commit "vendoring". Can we consolidate all the Go vendoring into 1 commit?

File: daemon/hostdaemon.go [hunk 12]
```diff
@@ -51,21 +56,37 @@ func NewHostDaemon(vsp plugin.VendorPlugin) *HostDaemon {
 	}
 }
 
+func (d *HostDaemon) ensureConnected() {
+	if d.conn != nil {
```
[the remark this answers is not in the document]
Comment (bn222) [reply daemon/hostdaemon.go:82]: When you `Dial` using `grpc`, the connection might not be up. For that reason, having an additional check here might not work. If there is no connection, it will error out later.

File: daemon/main.go [hunk 13]
```diff
@@ -29,7 +29,7 @@ func isDpuMode(mode string) (bool, error) {
 }
 
 func createDaemon(dpuMode bool) (Daemon, error) {
-	plugin := plugin.NewGrpcPlugin(dpuMode)
+	plugin := plugin.NewDummyPlugin() // plugin.NewGrpcPlugin(dpuMode)
```
[the remark this answers is not in the document]
Comment (bn222) [reply daemon/main.go]: Thanks for catching this.

File: vendor/github.com/golang/protobuf/jsonpb/decode.go
Comment (Eoghan1232) [inline vendor/github.com/golang/protobuf/jsonpb/decode.go:1]: can we remove vendor folder entirely?

File: bindata/daemon/04.daemonset.yaml [hunk 14]
```diff
@@ -26,13 +26,38 @@ spec:
...
         env:
         - name: K8S_NODE
           valueFrom:
             fieldRef:
               fieldPath: spec.nodeName
         - name: NAMESPACE
           value: {{.Namespace}}
+        volumeMounts:
+        - name: vendor-plugin-sock
+          mountPath: /var/run/dpu-daemon
```
Comment (Eoghan1232) [inline bindata/daemon/04.daemonset.yaml:39]: /var/run/daemon/dpu-daemon?

File: daemon/dpudaemon.go [hunk 15]
```diff
@@ -17,16 +17,17 @@ type DpuDaemon struct {
...
 	vsp plugin.VendorPlugin
 	log logr.Logger
+	server *grpc.Server
 }
 
 func (s *DpuDaemon) CreateBridgePort(context context.Context, bpr *pb.CreateBridgePortRequest) (*pb.BridgePort, error) {
-	s.log.Info("Create Bridge Port", "req", bpr.BridgePort.Name)
-	return &pb.BridgePort{}, nil
+	err := s.vsp.CreateBridgePort(bpr)
+	return nil, err
```
Comment (Eoghan1232) [inline daemon/dpudaemon.go:25]: why ``nil, err``?

File: daemon/hostdaemon.go [hunk 16]
```diff
@@ -58,14 +146,29 @@ func (d *HostDaemon) Start() {
...
+
+	add := func(r *cnitypes.PodRequest) (*cni100.Result, error) {
+		return d.addHandler(r)
+	}
+	del := func(r *cnitypes.PodRequest) (*cni100.Result, error) {
+		return d.delHandler(r)
+	}
+
+	server := cniserver.NewCNIServer(add, del)
+	err = server.Start()
```
Comment (Eoghan1232) [inline daemon/hostdaemon.go:160]: why is this here?

File: daemon/hostdaemon.go [hunk 17]
```diff
@@ -58,14 +146,29 @@ func (d *HostDaemon) Start() {
...
-	// listens to requests coming from the cni shim
-	select {}
+	d.cniserver = server
+	go func() {	
+		d.cniserver.Start()
+	}()
+}
+
+func (d *HostDaemon) Stop() {
+
```
Comment (Eoghan1232) [inline daemon/hostdaemon.go:173]:
```suggestion
	if d.cniserver != nil {
		d.cniserver.GracefulStop()
		d.cniserver = nil
	}
```

File: daemon/main.go [hunk 18]
```diff
@@ -37,6 +39,31 @@ func createDaemon(dpuMode bool) (Daemon, error) {
...
+	_, err = io.Copy(destinationFile, sourceFile)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func copyCNI() error {
+	return copyFile("/dpu-cni", "/opt/cni/bin/dpu-cni")
```
Comment (Eoghan1232) [inline daemon/main.go:64]: this should be part of an entrypoint script tbh, this is how sriov-cni does it
I don't think daemon should be responsible for copying the binary?

File: daemon/main.go (hunk 1 above)
Comment (Eoghan1232) [reply daemon/main.go:83]: +1

File: daemon/plugin/vendorplugin.go [hunk 19]
```diff
@@ -12,7 +12,7 @@ import (
 )
 
 const (
-	VendorPluginDir        string = "/var/run/daemon"
+	VendorPluginDir        string = "/var/run/daemon/vendor-plugin"
```
[the remark this answers is not in the document]
Comment (Eoghan1232) [reply daemon/plugin/vendorplugin.go:16]: /var/run/daemon/dpu-daemon should be used

File: daemon/plugin/vendorplugin.go [hunk 20]
```diff
@@ -3,50 +3,57 @@ package plugin
...
 	}
 }
 
-func (v *DummyPlugin) Start() (string, string) {
-	if runtime.GOARCH == "amd64" {
-		return "127.0.0.1", "50051"
-	} else {
-		return "127.0.0.1", "50051"
-	}
+func (v *DummyPlugin) Start() (string, int32, error) {
```
Comment (Eoghan1232) [inline daemon/plugin/vendorplugin.go:37]: we can probably remove this now

File: daemon/plugin/vendorplugin.go [hunk 21]
```diff
@@ -3,50 +3,57 @@ package plugin
...
+	return "127.0.0.1", 50051, nil
 }
 
 func (v *DummyPlugin) Stop() {
 
 }
 
+func (v *DummyPlugin) CreateBridgePort(createRequest *opi.CreateBridgePortRequest) error {
```
Comment (Eoghan1232) [inline daemon/plugin/vendorplugin.go:45]: same comment as above

File: daemon/plugin/vendorplugin.go [hunk 22]
```diff
@@ -59,6 +66,7 @@ func (g *GrpcPlugin) Start() (string, int32, error) {
 		g.log.Error(err, "Failed to start serving")
 		return "", 0, err
 	}
+	g.log.Info("Successful call to Init", ipPort.Ip, ipPort.Port)
```
Comment (Eoghan1232) [inline daemon/plugin/vendorplugin.go:69]: this is just a debug line right? remove?

File: daemon/plugin/vendorplugin.go [hunk 23]
```diff
@@ -94,5 +102,17 @@ func (g *GrpcPlugin) ensureConnected() error {
 	g.conn = conn
 
 	g.client = pb.NewLifeCycleServiceClient(conn)
+	g.opiClient = opi.NewBridgePortServiceClient(conn)
 	return nil
 }
+
+func (g *GrpcPlugin) CreateBridgePort(createRequest *opi.CreateBridgePortRequest) error {
+	g.ensureConnected()
+	_, err := g.opiClient.CreateBridgePort(context.TODO(), createRequest)
```
Comment (Eoghan1232) [inline daemon/plugin/vendorplugin.go:111]: this is where we will forward req to vsp right?

File: dpu-cni/pkgs/cniserver/cniserver.go [hunk 24]
```diff
@@ -24,17 +25,19 @@ import (
...
-	runDir         string
-	socketPath     string
+	addHandler processRequestFunc
+	delHandler processRequestFunc
+	socketPath string
 }
 
 // ensureRunDirExists makes sure that the socket being created is only accessible to root.
-func ensureRunDirExists(runDir string, socketPath string) error {
+func ensureRunDirExists(serverSocketPath string) error {
```
Comment (Eoghan1232) [inline dpu-cni/pkgs/cniserver/cniserver.go:37]: why is this needed?
on deletion should the socket not get cleaned up?

File: dpu-cni/pkgs/cniserver/cniserver.go (hunk 6 above)
Comment (Eoghan1232) [reply dpu-cni/pkgs/cniserver/cniserver.go:262]: +1

File: dpu-cni/pkgs/cniserver/cniserver.go [hunk 25]
```diff
@@ -319,15 +333,8 @@ func NewCNIServer(options ...func(*Server)) *Server {
...
 }
 
-func WithHandler(rqFunc processRequestFunc) func(*Server) {
-	return func(s *Server) {
-		s.processRequest = rqFunc
-	}
-}
-
-func WithSocketPath(runDir string, socketPath string) func(*Server) {
+func WithSocketPath(socketPath string) func(*Server) {
```
Comment (Eoghan1232) [inline dpu-cni/pkgs/cniserver/cniserver.go:336]: where is this func used?

Comment (Eoghan1232) [reply dpu-cni/pkgs/cniserver/cniserver.go:336]: also should the func be exported?

File: dpu-cni/pkgs/cnitypes/cnitypes.go [hunk 26]
```diff
@@ -10,9 +10,7 @@ import (
 	"github.com/vishvananda/netlink"
 )
 
-const ServerRunDir string = "/var/run/dpu-cni"
-const ServerSocketName string = "dpu-cni-server.sock"
-const ServerSocketPath string = ServerRunDir + "/" + ServerSocketName
+const ServerSocketPath string = "/var/run/dpu-daemon/dpu-cni/dpu-cni-server.sock"
```
[the remark this answers is not in the document]
Comment (Eoghan1232) [reply dpu-cni/pkgs/cnitypes/cnitypes.go:13]: +1
```suggestion
const ServerSocketPath string = "/var/run/dpu-daemon/dpu-cni/dpu-cni-server.sock"
const path string = "/var/run/dpu-daemon/"
const path2 string = "dpu-cni/dpu-cni-server.sock"
```

with actual names but it's handy to have 2 separated so we can construct new paths in future without refactoring

File: dpu-cni/pkgs/sriov/sriov.go [hunk 27]
```diff
@@ -198,9 +198,9 @@ func (s *sriovManager) ApplyVFConfig(conf *cnitypes.NetConf) error {
 		return fmt.Errorf("failed to lookup master %q: %v", conf.Master, err)
 	}
 	// 1. Set vlan
-	if err = s.nLink.LinkSetVfVlanQosProto(pfLink, conf.VFID, *conf.Vlan, *conf.VlanQoS, cnitypes.VlanProtoInt[*conf.VlanProto]); err != nil {
-		return fmt.Errorf("failed to set vf %d vlan configuration - id %d, qos %d and proto %s: %v", conf.VFID, *conf.Vlan, *conf.VlanQoS, *conf.VlanProto, err)
-	}
+	// if err = s.nLink.LinkSetVfVlanQosProto(pfLink, conf.VFID, *conf.Vlan, *conf.VlanQoS, cnitypes.VlanProtoInt[*conf.VlanProto]); err != nil {
```
[the remark this answers is not in the document]
Comment (Eoghan1232) [reply dpu-cni/pkgs/sriov/sriov.go:201]: no - this code is commented out right now due to driver not supporting this func.
I don't see a reason to create a flag based on support or not ? 

or you mean if another vendor supported and wanting this func  ?

File: dpu-cni/pkgs/cniserver/cniserver.go (hunk 25 above)
Comment (bn222) [reply dpu-cni/pkgs/cniserver/cniserver.go:336]: `dpu-cni/pkgs/cniserver/cniserver_test.go`

File: daemon/plugin/vendorplugin.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (bn222) [reply daemon/plugin/vendorplugin.go:16]: The intent is to have everything in `/var/run/dpu-daemon/`. The reason there is a subdirectory is that we need to set permissions on the directory.

For that reason, We should have:

`/var/run/dpu-daemon/vendor-plugin`
`/var/run/dpu-daemon/cni-server`
...

File: daemon/plugin/vendorplugin.go (hunk 20 above)
Comment (bn222) [reply daemon/plugin/vendorplugin.go:37]: I will move it to tests since that's where it's used and it's specific to testing.

File: daemon/main.go (hunk 1 above)
Comment (bn222) [reply daemon/main.go:83]: Can you elaborate on why we need to do the copy outside of the daemon?

File: vendor/github.com/golang/protobuf/jsonpb/decode.go
Comment (bn222) [reply vendor/github.com/golang/protobuf/jsonpb/decode.go:1]: We would like to but we can't.

File: internal/controller/dpuoperatorconfig_controller.go [hunk 28]
```diff
@@ -40,6 +40,12 @@ type DpuOperatorConfigReconciler struct {
 //+kubebuilder:rbac:groups=config.openshift.io,resources=dpuoperatorconfigs,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=config.openshift.io,resources=dpuoperatorconfigs/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=config.openshift.io,resources=dpuoperatorconfigs/finalizers,verbs=update
+//+kubebuilder:rbac:groups="",resources=serviceaccounts,verbs=get;list;watch;create;update;patch;delete
```
[the remark this answers is not in the document]
Comment (bn222) [reply internal/controller/dpuoperatorconfig_controller.go:44]: I didn't run any operator-sdk commands. I added this to tell kubebuilder about what I was missing.

File: daemon/main.go (hunk 18 above)
Comment (bn222) [reply daemon/main.go:64]: why not? What's the benefit of having an entrypoint script

File: dpu-cni/pkgs/sriov/sriov.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (bn222) [reply dpu-cni/pkgs/sriov/sriov.go:201]: I prefer to comment this out. It's a temporary workaround anyway.

File: daemon/plugin/vendorplugin.go (hunk 23 above)
Comment (bn222) [reply daemon/plugin/vendorplugin.go:111]: yes

File: daemon/plugin/vendorplugin.go (hunk 21 above)
Comment (bn222) [reply daemon/plugin/vendorplugin.go:45]: In case of CreateBridgePort in DummyPlugin (to be moved to test), I don't think we need to return an error. What would you want to return here?

Comment (Eoghan1232) [reply daemon/plugin/vendorplugin.go:45]: *pb.BridgePort  ?

File: daemon/plugin/vendorplugin.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (bn222) [reply daemon/plugin/vendorplugin.go:16]: What I propose is that we have `/var/run/dpu-daemon/` as a global var somewhere, and whenever we want to create a socket, we use concat it with `sub-dir/sock-name.socket`. We only need to pass around 1 variable. We can use `filepath.Dir` to get the directory and operate on that part separately.

File: daemon/main.go (hunk 1 above)
Comment (bn222) [reply daemon/main.go]: Took a note that we will need to address this. More detail on this: cp from rhel 9 can be used on rhel 8, but we can't use our go binary if it is running on rhel 9 container while the host system is still rhel 8

**Comment by bn222** [comment]
@SalDaniele PTAL

File: daemon/hostdaemon.go (hunk 3 above)
Comment (bn222) [reply daemon/hostdaemon.go]: Thanks for catching this. It's now resolved as part of the move to `ServeAndListen`.

File: daemon/plugin/vendorplugin.go [hunk 29]
```diff
@@ -54,6 +54,7 @@ func (g *GrpcPlugin) Start() (string, int32, error) {
 		g.log.Error(err, "Failed to start serving")
 		return "", 0, err
 	}
+	g.log.Info("Successful call to Init", ipPort.Ip, ipPort.Port)
```
[the remark this answers is not in the document]
Comment (bn222) [reply daemon/plugin/vendorplugin.go]: I've removed it. The assumption is that the vsp has proper tests

File: daemon/hostdaemon.go [hunk 30]
```diff
@@ -31,25 +32,33 @@ type HostDaemon struct {
...
-	_, err := d.client.CreateBridgePort(context.TODO(), createRequest)
+	_, err = d.client.CreateBridgePort(context.TODO(), createRequest)
 	return err
 }
 
 func (d *HostDaemon) DeleteBridgePort(pf int, vf int, vlan int, mac string) (error) {
 	d.ensureConnected()
 	req := &pb.DeleteBridgePortRequest{
-		    Name: fmt.Sprintf("%d-%d-%d-%s", pf, vf, vlan, mac),
+		Name: fmt.Sprintf("%d-%d-%d-%s", pf, vf, vlan, mac),
```
[the remark this answers is not in the document]
Comment (bn222) [reply daemon/hostdaemon.go:61]: The call to DeleteBridgePort isn't the same. I'm marking it as resolved since we don't need to change this right now but we might want to revisit this.

File: dpu-cni/pkgs/cniserver/cniserver.go (hunk 24 above)
Comment (bn222) [reply dpu-cni/pkgs/cniserver/cniserver.go]: No, we must ensure that we are dealing with a clean slate. You can't assume anything about the state of the filesystem.

File: dpu-cni/pkgs/cnitypes/cnitypes.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (bn222) [reply dpu-cni/pkgs/cnitypes/cnitypes.go]: see my other comment

File: go.mod (hunk 2 above)
Comment (vrindle) [reply go.mod:24]: Any chance we can split this up upgrades into seperate PRs? These changes to go.mod and go.sum make a lot of additional noise in the PR.

File: daemon/dpudaemon.go (hunk 15 above)
Comment (vrindle) [reply daemon/dpudaemon.go:25]: I have the same question. I have the same question, why are no longer returning an empty bridge port and instead of nil and err?  If we are only going to return nil, it would make sense to return only err.

File: daemon/dpudaemon.go [hunk 31]
```diff
@@ -21,13 +21,13 @@ type DpuDaemon struct {
...
 }
 
 func (s *DpuDaemon) DeleteBridgePort(context context.Context, bpr *pb.DeleteBridgePortRequest) (*emptypb.Empty, error) {
-	s.log.Info("Delete Bridge Port", "req", bpr.Name)
-	return nil, nil
+	err := s.vsp.DeleteBridgePort(bpr)
+	return nil, err
```
Comment (vrindle) [inline daemon/dpudaemon.go:30]: same comment as above

File: daemon/hostdaemon.go [hunk 32]
```diff
@@ -88,6 +107,43 @@ func (d *HostDaemon) connectWithRetry() {
 	d.client = pb.NewBridgePortServiceClient(conn)
 }
 
+func (d *HostDaemon) addHandler(req *cnitypes.PodRequest) (*cni100.Result, error) {
+	pf := 0
```
Comment (vrindle) [inline daemon/hostdaemon.go:111]: We don't want magic numbers in the codebase. Is it possible to make these as constants?

File: daemon/hostdaemon.go [hunk 33]
```diff
@@ -88,6 +107,43 @@ func (d *HostDaemon) connectWithRetry() {
...
+	vf := req.CNIConf.VFID
+	mac := req.CNIConf.MAC
+	vlan := 7
```
Comment (vrindle) [inline daemon/hostdaemon.go:114]: Same as comment above

File: daemon/hostdaemon.go [hunk 34]
```diff
@@ -88,6 +107,43 @@ func (d *HostDaemon) connectWithRetry() {
...
+	}
+	d.log.Info("addHandler sm.CmdAdd succeeded")
+	return res, nil
+}
+
+func (d *HostDaemon) delHandler(req *cnitypes.PodRequest) (*cni100.Result, error) {
+	pf := 0
+	vf := req.CNIConf.VFID
+	mac := req.CNIConf.MAC
+	vlan := 7
```
Comment (vrindle) [inline daemon/hostdaemon.go:135]: We are reusing these. They should definitely be in a const.

File: daemon/hostdaemon.go (hunk 32 above)
Comment (Eoghan1232) [reply daemon/hostdaemon.go:111]: these are dummy values for testing and will be removed.
the actual values will come from the req.

File: daemon/hostdaemon.go (hunk 34 above)
Comment (Eoghan1232) [reply daemon/hostdaemon.go:135]: same as above.

@bn222 can you remove these and use values from the req instead?

**Comment by Eoghan1232** [comment]
daemon/dpudaemon.go
I needed to add a `select{}` at the end of the Start() func - so the daemon would not exit out and complete afterwards

@bn222 we know about this, but it never got added to the PR I guess.

File: daemon/main_test.go [hunk 35]
```diff
@@ -59,7 +59,7 @@ var _ = g.Describe("Main", func() {
...
 			HostDaemon.Start()
 			DpuDaemon.Start()
-			_, err := HostDaemon.CreateBridgePort(1, 1, 1)
+			_, err := HostDaemon.CreateBridgePort(1, 1, 1, "00:11:22:33:44:55")
```
Comment (wizhaoredhat) [inline daemon/main_test.go]: Squash this with the previous commit.

File: daemon/hostdaemon.go [hunk 36]
```diff
@@ -34,23 +34,36 @@ type HostDaemon struct {
...
-			Spec: &pb.BridgePortSpec{
-				Ptype:      1,
-				MacAddress: macBytes,
-				LogicalBridges: []string{
-					fmt.Sprintf("%d", vlan),
-				},
-			},
-		},
-	}
+    BridgePort: &pb.BridgePort{
```
Comment (wizhaoredhat) [inline daemon/hostdaemon.go]: Indentation issues.

File: daemon/hostdaemon.go [hunk 37]
```diff
@@ -34,23 +34,36 @@ type HostDaemon struct {
...
+        Name: "m" + fmt.Sprint(1),
```
Comment (wizhaoredhat) [inline daemon/hostdaemon.go]: I am confused what this is...

File: daemon/hostdaemon.go [hunk 38]
```diff
@@ -34,23 +34,36 @@ type HostDaemon struct {
...
+        Spec: &pb.BridgePortSpec{
+            Ptype:          1,
+            MacAddress:     m,
+            LogicalBridges: []string{fmt.Sprint(1 + 2)},
```
Comment (wizhaoredhat) [inline daemon/hostdaemon.go]: Missing VLAN and PF/VLAN info???

Comment (wizhaoredhat) [reply daemon/hostdaemon.go]: What is wrong with using the commented out code from below?

File: daemon/hostdaemon.go [hunk 39]
```diff
@@ -161,7 +174,7 @@ func (d *HostDaemon) Start() {
 		return d.delHandler(r)
 	}
 
-	d.cniserver = cniserver.NewCNIServer(add, del)
+	d.cniserver = cniserver.NewCNIServer(add, del, cniserver.WithSocketPath(d.cniServerPath))
```
Comment (wizhaoredhat) [inline daemon/hostdaemon.go:188]: I think this code change should be squash with the first commit in this series for "each commit" should compile.

File: daemon/hostdaemon.go [hunk 40]
```diff
@@ -63,7 +67,6 @@ func (d *HostDaemon) CreateBridgePort(pf int, vf int, vlan int, mac string) (*pb
 //			},
 //		},
 //	}
-	print("------")
```
Comment (wizhaoredhat) [inline daemon/hostdaemon.go]: squash with previous commit

File: daemon/main_test.go [hunk 41]
```diff
@@ -45,6 +46,36 @@ func (v *DummyPlugin) DeleteBridgePort(deleteRequest *opi.DeleteBridgePortReques
 	return nil
 }
 
+type SriovManagerMock struct {}
```
Comment (wizhaoredhat) [inline daemon/main_test.go:49]: Just a general question whether we will be using Mocks for HW functionality?

**Review by wizhaoredhat — CHANGES_REQUESTED** [review]

File: daemon/main_test.go (hunk 41 above)
Comment (bn222) [reply daemon/main_test.go:49]: We should use a stub (or ideally a fake) when dealing with deeper layers of the architecture. We shouldn't blindly mock everything that goes to the hardware. Rather, we should write code that encapsulates behavior.

File: daemon/main_test.go [hunk 42]
```diff
@@ -49,18 +115,58 @@ var _ = g.BeforeSuite(func() {
...
+			Expect(err).NotTo(HaveOccurred())
+			go func() {
+				err = dpuDaemon.Serve(dpuListen)
+				Expect(err).NotTo(HaveOccurred())
+			}()
+
+			hostListen, err := hostDaemon.Listen()
+			Expect(err).NotTo(HaveOccurred())
+			go func() {
+				_ = hostDaemon.Serve(hostListen)
```
Comment (wizhaoredhat) [inline daemon/main_test.go:158]: Same as above we should check the return `err`?

File: daemon/main_test.go [hunk 43]
```diff
@@ -49,18 +115,58 @@ var _ = g.BeforeSuite(func() {
...
+		hostDaemon.Stop()
+	})
 
-		g.It("Should connect succesfully if the server is up first", func() {
-			HostDaemon.Start()
-			DpuDaemon.Start()
-			_, err := HostDaemon.CreateBridgePort(1, 1, 1)
-			Expect(err).ShouldNot(HaveOccurred())
+	g.Context("Connection", func() {
+		g.It("Should connect successfully if the DPU Daemon is up first", func() {
```
Comment (wizhaoredhat) [inline daemon/main_test.go:147]: Shouldn't we have another test case to see if Host Daemon is started first?

File: daemon/hostdaemon.go [hunk 44]
```diff
@@ -82,18 +112,96 @@ func (d *HostDaemon) connectWithRetry() {
...
+
+	if err != nil {
+		d.log.Error(err, "Failed to listen")
+		return err
+	}
+	return d.Serve(listener)
+}
+
+func (d *HostDaemon) Serve(listner net.Listener) error {
+	err := d.cniserver.Serve(listner)
```
Comment (wizhaoredhat) [inline daemon/hostdaemon.go:192]: SP: listener

File: daemon/hostdaemon.go [hunk 45]
```diff
@@ -82,18 +112,96 @@ func (d *HostDaemon) connectWithRetry() {
...
+	}
+	d.log.Info("addHandler d.sm.CmdAdd succeeded")
+	return res, nil
+}
+
+func (d *HostDaemon) cniCmdDelHandler(req *cnitypes.PodRequest) (*cni100.Result, error) {
+	pf := 0
+	vf := req.CNIConf.VFID
+	mac := req.CNIConf.MAC
+	vlan := 7//*req.CNIConf.Vlan
```
Comment (wizhaoredhat) [inline daemon/hostdaemon.go:148]: We should just use req.CNIConf.Vlan here. I don't see why this won't work.

File: daemon/hostdaemon.go [hunk 46]
```diff
@@ -82,18 +112,96 @@ func (d *HostDaemon) connectWithRetry() {
...
 	d.conn = conn
 	d.client = pb.NewBridgePortServiceClient(conn)
+	return nil
+}
+
+func (d *HostDaemon) cniCmdAddHandler(req *cnitypes.PodRequest) (*cni100.Result, error) {
+	pf := 0
+	vf := req.CNIConf.VFID
+	mac := req.CNIConf.RuntimeConfig.Mac
+	d.log.Info("addHandler", "CNIConf", req.CNIConf)
```
Comment (wizhaoredhat) [inline daemon/hostdaemon.go:126]: I think we can remove these prints.

**Review by wizhaoredhat — CHANGES_REQUESTED** [review]
Overall looks good after squashing.

File: daemon/main_test.go (hunk 42 above)
Comment (bn222) [reply daemon/main_test.go:158]: We should not because serve can fail and that's expected.

File: daemon/main_test.go (hunk 43 above)
Comment (bn222) [reply daemon/main_test.go:147]: No, not today. In the future, we might want to abstract out using a robust way to communicate. It shouldn't be tested at this level imo.

File: daemon/hostdaemon.go (hunk 45 above)
Comment (bn222) [reply daemon/hostdaemon.go:148]: thanks for catching this commented out code.

File: daemon/hostdaemon.go (hunk 46 above)
Comment (bn222) [reply daemon/hostdaemon.go:126]: Are you sure? last time I removed logs, the first thing we ended up doing during debug sessions is put the logs back.

File: daemon/main_test.go (hunk 42 above)
Comment (wizhaoredhat) [reply daemon/main_test.go:158]: But serve is not allowed to fail for the DPU Daemon?

File: daemon/main_test.go (hunk 43 above)
Comment (wizhaoredhat) [reply daemon/main_test.go:147]: What is the purpose of the test, just to make sure the DPU/Host Daemon can talk to each other?

File: daemon/hostdaemon.go (hunk 46 above)
Comment (wizhaoredhat) [reply daemon/hostdaemon.go:126]: I think the VLAN one is redundant. But very true for CNIConf

File: daemon/main_test.go (hunk 43 above)
Comment (bn222) [reply daemon/main_test.go:147]: This tests the fact that CNI calls get processed correctly. In the future, we will inject a dependency into HostDaemon to test the cases where we do plumbing.

**Comment by bn222** [comment]
@wizhaoredhat ptal
